File: vision/detect.js
// vision/detect.js
import cv from 'opencv4nodejs';
export function mapRoi(roi, shotW, shotH, viewportW, viewportH) {
  const sx = Math.round(roi.x * (shotW / Math.max(1, viewportW)));
  const sy = Math.round(roi.y * (shotH / Math.max(1, viewportH)));
  const sw = Math.round(roi.w * (shotW / Math.max(1, viewportW)));
  const sh = Math.round(roi.h * (shotH / Math.max(1, viewportH)));
  // clamp
  return new cv.Rect(
    Math.max(0, Math.min(sx, shotW - 1)),
    Math.max(0, Math.min(sy, shotH - 1)),
    Math.max(1, Math.min(sw, shotW - Math.max(0, Math.min(sx, shotW - 1)))),
    Math.max(1, Math.min(sh, shotH - Math.max(0, Math.min(sy, shotH - 1))))
  );
}

File: vision/detect.py
# vision/detect.py

def map_roi(roi, shot_w, shot_h, viewport_w, viewport_h):
    sx = round(roi['x'] * (shot_w / max(1, viewport_w)))
    sy = round(roi['y'] * (shot_h / max(1, viewport_h)))
    sw = round(roi['w'] * (shot_w / max(1, viewport_w)))
    sh = round(roi['h'] * (shot_h / max(1, viewport_h)))
    x = max(0, min(sx, shot_w - 1))
    y = max(0, min(sy, shot_h - 1))
    w = max(1, min(sw, shot_w - x))
    h = max(1, min(sh, shot_h - y))
    return (x, y, w, h)
